cpp/src/io/parquet/decode_fixed.cu

```diff
@@ -0,0 +1,507 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "page_data.cuh"
+#include "page_decode.cuh"
+#include "parquet_gpu.hpp"
+#include "rle_stream.cuh"
+
+#include <cudf/detail/utilities/cuda.cuh>
+
+namespace cudf::io::parquet::detail {
+
+namespace {
+
+constexpr int decode_block_size = 128;
+constexpr int rolling_buf_size  = decode_block_size * 2;
+// the required number of runs in shared memory we will need to provide the
+// rle_stream object
+constexpr int rle_run_buffer_size = rle_stream_required_run_buffer_size<decode_block_size>();
+
+template <bool nullable, typename level_t, typename state_buf>
+static __device__ int gpuUpdateValidityOffsetsAndRowIndicesFlat(int32_t target_value_count,
+                                                                page_state_s* s,
+                                                                state_buf* sb,
+                                                                level_t const* const def,
+                                                                int t,
+                                                                bool has_no_nulls)
+{
+  constexpr int num_warps      = decode_block_size / cudf::detail::warp_size;
+  constexpr int max_batch_size = num_warps * cudf::detail::warp_size;
+
+  auto& ni = s->nesting_info[0];
+
+  // how many (input) values we've processed in the page so far
+  int value_count = s->input_value_count;
+  int valid_count = ni.valid_count;
+
+  // cap by last row so that we don't process any rows past what we want to output.
+  int const first_row = s->first_row;
+  int const last_row  = first_row + s->num_rows;
+  target_value_count  = min(target_value_count, last_row);
+
+  int const valid_map_offset      = ni.valid_map_offset;
+  int const row_index_lower_bound = s->row_index_lower_bound;
+
+  __syncthreads();
```

Comment:
why is this sync needed?

Reply:
This function was already almost in this form in Dave's PR, so will ping him here for :eyes: @nvdbaranec. But, I don't see other `__syncthreads()` in this function when the page is not `nullable` and we are reading `valid_count` and `value_count` from shared memory and `t == 0` could advance towards the end of the function where it updates the shared memory values. I believe that means `t > 0` could get a different value of these shared mem variables, leading to issues because other threads would think they are further ahead and they should be.